cpp/src/heap.cuh

```diff
@@ -0,0 +1,225 @@
+       *   the tree.
+       *
+       * @param[in]  index - the current array index
+       * @return     the index of the parent of the current index
+       */
+      template <typename S>
+      inline S __host__ __device__ parent(S index) {
+	static_assert(std::is_integral<S>::value, "Index must be of an integral type");
+      
+	return ((index + 1) / 2) - 1;
+      }
+
+      /**
+       * @brief Identify the left child index of the specified index.
+       *        NOTE: This function does no bounds checking, so
+       *        the left child computed might be out of bounds.
+       *
+       *   See the above documentation for a picture to describe
+       *   the tree.
+       *
+       * @param[in]  index - the current array index
+       * @return     the index of the left child of the current index
+       */
+      template <typename S>
+      inline S __host__ __device__ left_child(S index) {
+	static_assert(std::is_integral<S>::value, "Index must be of an integral type");
+      
+	return ((index + 1) * 2 - 1);
+      }
+
+      /**
+       * @brief Identify the right child index of the specified index.
+       *        NOTE: This function does no bounds checking, so
+       *        the right child computed might be out of bounds.
+       *
+       *   See the above documentation for a picture to describe
+       *   the tree.
+       *
+       * @param[in]  index - the current array index
+       * @return     the index of the right child of the current index
+       */
+      template <typename S>
+      inline S __host__ __device__ right_child(S index) {
+	static_assert(std::is_integral<S>::value, "Index must be of an integral type");
+      
+	return (index + 1) * 2;
+      }
+    }
+      
+    /**
+     * @brief Reorder an existing array of elements into a heap
+     *
+     * @param[in, out]   array   - the existing array
+     * @param[in]        size    - the number of elements in the existing array
+     * @param[in]        compare - the comparison function to use
+     *
+     */
+    template <typename T, typename S, typename C>
+    inline void __host__ __device__ heapify(T *array, S size, C compare) {
+      static_assert(std::is_integral<S>::value, "Index must be of an integral type");
+
+      //
+      // We want to order ourselves as a heap.  This is accomplished by starting
+      // at the end and for each element, compare with its parent and
+      // swap if necessary.  We repeat this until there are no more swaps
+      // (should take no more than log2(size) iterations).
+      //
+      S count_swaps = 1;
+      while (count_swaps > 0) {
+	count_swaps = 0;
+	for (S i = size - 1 ; i > 0 ; --i) {
+	  S p = detail::parent(i);
+
+	  {
+	    S lc = detail::left_child(p);
+	    S rc = detail::right_child(p);
+	    if (i == lc) {
+	    } else if (i == rc) {
+	    } else {
+	      printf("in heapify... i = %ld, p = %ld, lc = %ld, rc = %ld\n", i, p, lc, rc);
```

Comment:
Are we going to keep this printf? I guess this is for debugging and cuGraph users don't need this information.

Reply:
They were for debugging... forgot they were still there.  I'll clean them up.